File: sdk/dxn3.js
/**
 * dxn3 SDK — write a game in JavaScript (node), the studio renders it.
 *
 *   const { rect, label, on, run, dt } = require("dxn3");  // NODE_PATH=sdk
 *
 *   const ship = rect("ship", W >> 1, H - 12, 12, 5, "#8b5cf6");
 *   ship.tag = "ship";
 *
 *   on.key(k => { if (k === "left") ship.x -= 340 * dt; });
 *   on.tick(() => { ... });            // every frame — read dt live
 *   on.hit((a, b) => { ... });         // two tagged entities just overlapped
 *   run();
 *
 * The engine owns rendering, input and collision; this module speaks the
 * line-JSON protocol on stdio. console.log goes to the studio's console.
 * (plain CJS on purpose — the studio's gates keep package.json out of the
 *  tree, and NODE_PATH makes `require('dxn3')` work from anywhere.)
 */
"use strict";

const fs = require("fs");

function _readLineSync() {
  const buf = Buffer.alloc(1);
  let out = "";
  for (;;) {
    const n = fs.readSync(0, buf, 0, 1, null);
    if (n === 0) return out.length ? out : null;   // EOF — the engine left
    if (buf[0] === 0x0a) return out;
    out += String.fromCharCode(buf[0]);
  }
}

// the engine says hello before your module finishes — W and H are ready
// for your very first line of code
const _hello = _readLineSync();
if (_hello === null) {
  console.log("run me from the studio:  dxn3 mygame.js   (not directly)");
  process.exit(1);
}
let W = 100, H = 46;
try { const h = JSON.parse(_hello); W = h.w | 0 || 100; H = h.h | 0 || 46; }
catch { /* defaults hold */ }

let dt = 0.0;                                    // read dt live, every frame
const E = new Map();                             // name -> entity object
const _scene = { t: "scene", name: "game", bg: "#0b0e1a", gravity: 0, entities: [] };
let _dels = [];
let _vars = {};
let _cam = {};
const _cb = { tick: [], key: [], hit: [], start: [] };

function _send(o) { process.stdout.write(JSON.stringify(o) + "\n"); }

function _mk(name, d) {
  d.visible = 1;
  if (E.has(name)) {
    // same name again = a redraw: replace the body in place
    const i = _scene.entities.findIndex((e) => e.name === name);
    if (i >= 0) _scene.entities[i] = d;
  } else {
    _scene.entities.push(d);
  }
  E.set(name, d);
  return d;
}
function rect(name, x, y, w, h, color = "#8b5cf6") {
  return _mk(name, { name, shape: "rect", x, y, w, h, color });
}
function circle(name, x, y, w, h, color = "#facc15") {
  return _mk(name, { name, shape: "circle", x, y, w, h, color });
}
function tri(name, x, y, w, h, color = "#ef4444") {
  return _mk(name, { name, shape: "tri", x, y, w, h, color });
}
function label(name, x, y, text, color = "#e9e5ff") {
  return _mk(name, { name, shape: "text", x, y,
                     w: Math.max(1, String(text).length), h: 2,
                     text: String(text), color });
}
function destroy(name) { if (E.has(name)) { E.delete(name); _dels.push(name); } }
function find(name) { return E.get(name); }
function background(c) { _scene.bg = c; }
function gravity(g) { _scene.gravity = g; }
function magnet(m) { _scene.magnet = m; }
function camera(x, y, zoom = 1) { _cam = { x, y, zoom }; }
function vars(kv) { Object.assign(_vars, kv); }
const on = {
  tick: fn => _cb.tick.push(fn),
  key: fn => _cb.key.push(fn),
  hit: fn => _cb.hit.push(fn),
  start: fn => _cb.start.push(fn),
};

function run() {
  _send(_scene);
  let started = false;
  let pairs = new Set();                 // overlap pairs already reported
  for (;;) {
    const line = _readLineSync();
    if (line === null) break;            // the studio closed the pipe
    let pkt;
    try { pkt = JSON.parse(line); }
    catch { console.log(line.slice(0, 200)); continue; }  // chatter -> console
    if (pkt.t !== "tick") continue;
    dt = +pkt.dt || 0;
    const keys = pkt.keys || {};
    if (!started) { _cb.start.forEach(f => f()); started = true; }
    for (const e of E.values()) {        // physics-lite
      if (e.vx) e.x += e.vx * dt;
      if (e.vy) e.y += e.vy * dt;
    }
    _cb.tick.forEach(f => f(dt));
    const held = ["left", "right", "jump", "space"].filter(k => keys[k]);
    held.push(...String(pkt.chars || "").split("").filter(c => c.trim()));
    for (const k of held) _cb.key.forEach(f => f(k));
    // hits arrive flat: [nameA, nameB, …]. fire on ENTER only — a pair
    // that separates may fire again; a held overlap never spams
    const hs = Array.isArray(pkt.hits) ? pkt.hits : [];
    const seen = new Set();
    for (let i = 0; i + 1 < hs.length; i += 2) {
      const na = String(hs[i]), nb = String(hs[i + 1]);
      const pair = na <= nb ? na + "\u0000" + nb : nb + "\u0000" + na;
      seen.add(pair);
      if (pairs.has(pair)) continue;
      const a = E.get(na), b = E.get(nb);
      if (a && b) _cb.hit.forEach(f => f(a, b));
    }
    pairs = seen;
    _send({ t: "frame", set: [...E.values()], del: _dels,
            vars: _vars, camera: _cam });
    _dels = []; _vars = {};
  }
}

module.exports = { W, H, get dt() { return dt; },
                   rect, circle, tri, label, destroy, find,
                   background, gravity, magnet, camera, vars, on, run };

File: sdk/dxn3.py
"""dxn3 SDK — write a game in Python, the studio renders it.

    from dxn3 import *
    import random

    ship = rect("ship", W // 2, H - 10, 12, 5, "#8b5cf6")

    def on_key(k):          # held keys each frame: "left" "right" "jump" "space" + letters
        if k == "left": ship.x -= 1

    def on_tick(dt):        # every frame
        ...

    def on_hit(a, b):       # two tagged entities overlapping
        ...

    run()                   # hands the loop to the studio

The engine owns rendering, input and collision; this module talks the
line-JSON protocol on stdio. print() goes to the studio's console.
"""
import json
import sys

# the engine says hello before your module finishes importing — that is
# how W and H are ready for your very first line of code
try:
    _hello = json.loads(sys.stdin.readline())
    W = int(_hello.get("w", 100))
    H = int(_hello.get("h", 46))
except Exception:
    print("run me from the studio:  dxn3 mygame.py   (not directly)")
    sys.exit(1)

dt = 0.0                                    # seconds since last tick
E = {}                                      # name -> Ent
_scene = {"t": "scene", "name": "game", "bg": "#0b0e1a",
          "gravity": 0, "entities": []}
_dels = []
_vars = {}
_cam = {}
_ticks, _keys, _hits, _starts = [], [], [], []


class Ent:
    """one entity — attributes ARE the fields the engine renders."""
    def __init__(self, d):
        object.__setattr__(self, "d", d)
    def __getattr__(self, k):
        return self.d.get(k)
    def __setattr__(self, k, v):
        self.d[k] = v


def _send(o):
    sys.stdout.write(json.dumps(o) + "\n")
    sys.stdout.flush()


def _mk(name, d):
    d.setdefault("visible", 1)
    if name in E:
        # same name again = a redraw: replace the body in place (both in
        # the registry and the pending scene packet) instead of raising
        for i, old in enumerate(_scene["entities"]):
            if old.get("name") == name:
                _scene["entities"][i] = d
                break
    else:
        _scene["entities"].append(d)
    ent = Ent(d)
    E[name] = ent
    return ent


def rect(name, x, y, w, h, color="#8b5cf6"):
    return _mk(name, {"name": name, "shape": "rect", "x": x, "y": y,
                      "w": w, "h": h, "color": color})


def circle(name, x, y, w, h, color="#facc15"):
    return _mk(name, {"name": name, "shape": "circle", "x": x, "y": y,
                      "w": w, "h": h, "color": color})


def tri(name, x, y, w, h, color="#ef4444"):
    return _mk(name, {"name": name, "shape": "tri", "x": x, "y": y,
                      "w": w, "h": h, "color": color})


def label(name, x, y, text, color="#e9e5ff"):
    return _mk(name, {"name": name, "shape": "text", "x": x, "y": y,
                      "w": max(1, len(str(text))), "h": 2, "text": str(text),
                      "color": color})


def destroy(name):
    if name in E:
        del E[name]
        _dels.append(name)


def find(name):
    return E.get(name)


def background(hex_color):
    _scene["bg"] = hex_color


def gravity(g):
    _scene["gravity"] = g


def magnet(px_):
    _scene["magnet"] = px_


def camera(x, y, zoom=1):
    _cam.update(x=x, y=y, zoom=zoom)


def vars(**kw):
    _vars.update(kw)


def on_tick(fn): _ticks.append(fn); return fn
def on_key(fn): _keys.append(fn); return fn
def on_hit(fn): _hits.append(fn); return fn
def on_start(fn): _starts.append(fn); return fn


def run():
    """hand the loop to the studio. your plain `def on_tick(dt)`,
    `def on_key(k)`, `def on_hit(a, b)`, `def on_start()` are picked
    up automatically — no decorators, no ceremony."""
    global dt
    import inspect
    fg = inspect.stack()[1].frame.f_globals
    def _hook(name):
        fn = fg.get(name)                       # skip our own registrars
        return fn if callable(fn) and getattr(fn, "__module__", "") != __name__ else None
    ticks = _ticks + ([_hook("on_tick")] if _hook("on_tick") else [])
    keys_ = _keys + ([_hook("on_key")] if _hook("on_key") else [])
    hits_ = _hits + ([_hook("on_hit")] if _hook("on_hit") else [])
    starts = _starts + ([_hook("on_start")] if _hook("on_start") else [])
    _send(_scene)
    started = False
    _pairs = set()                             # overlap pairs already reported
    for line in sys.stdin:
        try:
            pkt = json.loads(line)
        except Exception:
            print(line[:200])               # engine chatter -> console
            continue
        if pkt.get("t") != "tick":
            continue
        dt = float(pkt.get("dt") or 0.0)
        fg["dt"] = dt               # the caller's dt stays fresh — a copied
        keys = pkt.get("keys", {})   # float from `import *` would freeze at 0
        if not started:
            for fn in starts:
                fn()
            started = True
        # physics-lite: velocities integrate unless you steer them
        for ent in list(E.values()):
            vx = ent.d.get("vx", 0)
            vy = ent.d.get("vy", 0)
            if vx:
                ent.d["x"] = ent.d.get("x", 0) + vx * dt
            if vy:
                ent.d["y"] = ent.d.get("y", 0) + vy * dt
        for fn in ticks:
            fn(dt)
        held = [k for k in ("left", "right", "jump", "space") if keys.get(k)]
        held += [c for c in str(pkt.get("chars", "")) if c.strip()]
        for k in held:
            for fn in keys_:
                fn(k)
        # hits arrive flat: [nameA, nameB, nameC, …]. fire on ENTER only —
        # a pair that separates may fire again; a held overlap never spams
        hs = pkt.get("hits", []) or []
        seen = set()
        for i in range(0, len(hs) - 1, 2):
            na, nb = str(hs[i]), str(hs[i + 1])
            pair = (na, nb) if na <= nb else (nb, na)
            seen.add(pair)
            if pair in _pairs:
                continue
            a = E.get(na)
            b = E.get(nb)
            if a is not None and b is not None:
                for fn in hits_:
                    fn(a, b)
        _pairs = seen
        _send({"t": "frame",
               "set": [e.d for e in E.values()],
               "del": _dels,
               "vars": _vars,
               "camera": _cam})
        _dels.clear()
        _vars.clear()
